lightning/src/ln/outbound_payment.rs

```diff
@@ -455,7 +459,46 @@ impl RecipientOnionFields {
 	/// [`ChannelManager::send_spontaneous_payment`]: super::channelmanager::ChannelManager::send_spontaneous_payment
 	/// [`RecipientOnionFields::secret_only`]: RecipientOnionFields::secret_only
 	pub fn spontaneous_empty() -> Self {
-		Self { payment_secret: None, payment_metadata: None }
+		Self { payment_secret: None, payment_metadata: None, custom_tlvs: Vec::new() }
+	}
+
+	/// Creates a new [`RecipientOnionFields`] from an existing one, adding custom TLVs. Each
+	/// TLV is provided as a `(u64, Vec<u8>)` for the type number and serialized value
+	/// respectively. TLV type numbers must be unique, in increasing order, and within the range
+	/// reserved for custom types, i.e. >= 2^16, otherwise this method will return `Err(())`.
+	///
+	/// This method will also error for types in the experimental range which have been
+	/// standardized within the protocol, which only includes 5482373484 (keysend) for now.
+	///
+	/// See [`Self::custom_tlvs`] for more info.
+	pub fn with_custom_tlvs(mut self, mut custom_tlvs: Vec<(u64, Vec<u8>)>) -> Result<Self, ()> {
+		custom_tlvs.sort_unstable_by_key(|(typ, _)| *typ);
+		let mut prev_type = None;
+		for (typ, _) in custom_tlvs.iter() {
+			if *typ < 1 << 16 { return Err(()); }
+			if *typ == 5482373484 { return Err(()); } // keysend
+			match prev_type {
+				Some(prev) if prev >= *typ => return Err(()),
+				_ => {},
+			}
+			prev_type = Some(*typ);
+		}
+		self.custom_tlvs = custom_tlvs;
+		Ok(self)
+	}
+
+	/// Gets the custom TLVs that will be sent or have been received.
```

Comment:
Maybe the mention “have been received” can be dropped as this is method is about custom HTLC TLVs that are yet to be sent as part of `RecipientOnionField`.

Reply:
I would note the structure is called `RecipientOnionFields` with the following comment “Information which is provided, encrypted, to the payment recipient when sending HTLCs”.